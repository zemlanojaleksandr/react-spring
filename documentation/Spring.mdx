---
name: Spring
route: /spring
order: 7
---

import { Playground, PropsTable } from 'docz'
import { Spring, animated, config } from '../src/targets/web'
import { AnimatedMeter, RewindSpringProvider, RewindSpring } from './'
import { UnControlled as CodeMirror } from 'react-codemirror2'
import { Helmet } from 'react-helmet'

<Helmet>
  <style type="text/css">{`
    .CodeMirror {
      font-size: 14px;
      height: 100%;
    }
    .CodeMirror pre {
      line-height: 1.8!important;
    }
    .CodeMirror-line {
      padding: 0 10px!important;
    }
    .CodeMirror-lines {
      padding: 10px 0!important;
    }
    .CodeMirror-linenumber {
      padding: 0 7px 0 5px!important;
    }
    .code-table {
      display: flex;
    }
    .code-table > div {
      flex: 1;
    }
    .code-table > div:last-child {
      position: relative;
      display: flex;
      align-items: center;
      justify-content: center;
      background: #fb467b;
      font-family: -apple-system, BlinkMacSystemFont, avenir next, avenir, helvetica neue, helvetica, ubuntu, roboto, noto,
    segoe ui, arial, sans-serif;
      font-size: 4em;
      font-weight: 600;
      color: white;
    }
    .highlight {
      background: #f4e6e9;
      color: #947679;
      margin: 0 3px;
      padding: 4px 6px;
      border-radius: 3px;
    }
    .grommetux-meter {
      height: 70px;
    }
    .grommetux-meter__graphic {
      fill: transparent;
      stroke-linecap: round;
      stroke-linejoin: round;
      stroke-width: 15px;
      width: 70px;
      height: 70px;
    }
    .grommetux-meter__tracks {
      stroke: rgba(255,255,255,0.1);
    }
    .grommetux-meter__values {
      stroke: white;
    }
    code {
      background: #f4f6f9;
      color: #7D899C;
      margin: 0 3px;
      padding: 4px 6px;
      border-radius: 3px;
      font-family: "Source Code Pro",monospace;
      font-size: 14px;
    }
  `}</style>
</Helmet>

# Spring

You begin by importing `Spring`.

```jsx
import { Spring } from 'react-spring'
```

<RewindSpringProvider>
<p>
  Now, the primary task of a <code>Spring</code> is to move data <code>from</code> one state <code>to</code> another. <span className="highlight">Do not think of these values as "styles"</span> per se, although you can most certainly use them as such.
</p>

<div className="code-table">
  <CodeMirror
    value={
`<Spring
  from={{ opacity: 0 }}
  to={{ opacity: 1 }}>
  {props => <div style={props}>‚úåÔ∏è</div>}
</Spring>`}
    options={{
      mode: 'jsx',
      theme: 'dracula',
      lineNumbers: true,
      readOnly: true,
      showCursorWhenSelecting: false,
      cursorHeight: 0
    }}
  />
  <div>
    <RewindSpring>
      {x => <animated.div style={{ opacity: x }}>‚úåÔ∏è</animated.div>}
    </RewindSpring>
  </div>
</div>

<p>
  Do with them whatever you like, animate <span className="highlight">attributes</span> for instance, 
</p>

<div className="code-table">
  <CodeMirror
    value={
`<Spring
  from={{ x: 100 }}
  to={{ x: 0 }}>
  {props => (
    <svg strokeDashoffset={props.x}>
      <path d="..." />
    </svg>
  )}
</Spring>`}
    options={{
      mode: 'jsx',
      theme: 'dracula',
      lineNumbers: true,
      readOnly: true,
      showCursorWhenSelecting: false,
      cursorHeight: 0
    }}
  />
  <div>
    <RewindSpring>
      {x => (
        <animated.svg
          style={{ width: 100, height: 100, opacity: x }}
          viewBox="0 0 24 24"
          fill="none"
          stroke="white"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeDasharray={100}
          strokeDashoffset={x.interpolate(x => (1-x) * 100)}
        >
          <title>task-checklist-check</title>
          <path
            pathLength="100"
            className="a"
            d="M10,23.5H1.5a1,1,0,0,1-1-1V3.5a1,1,0,0,1,1-1h2"
          />
          <path
            pathLength="100"
            className="a"
            d="M15.5,2.5h2a1,1,0,0,1,1,1v6"
          />
          <path
            pathLength="100"
            className="a"
            d="M11.5,2.5a2,2,0,0,0-4,0h-2v3h8v-3Z"
          />
          <path
            pathLength="100"
            className="a"
            d="M16.5,9.5V5a.5.5,0,0,0-.5-.5H13.5"
          />
          <path
            pathLength="100"
            className="a"
            d="M5.5,4.5H3a.5.5,0,0,0-.5.5V21a.5.5,0,0,0,.5.5h7"
          />
          <line
            pathLength="100"
            className="a"
            x1="5.5"
            y1="8.499"
            x2="13.5"
            y2="8.499"
          />
          <line
            pathLength="100"
            className="a"
            x1="5.5"
            y1="11.499"
            x2="11.5"
            y2="11.499"
          />
          <line
            pathLength="100"
            className="a"
            x1="5.5"
            y1="14.499"
            x2="9.5"
            y2="14.499"
          />
          <circle pathLength="100" className="a" cx="17.5" cy="17.499" r="6" />
          <path
            pathLength="100"
            className="a"
            d="M20.174,15.755l-2.905,3.874a.75.75,0,0,1-1.13.08l-1.5-1.5"
          />
        </animated.svg>
      )}
    </RewindSpring>
  </div>
</div>

<p>
  or <span className="highlight">innerText</span>,
</p>

<div className="code-table">
  <CodeMirror
    value={
`<Spring
  from={{ number: 0 }}
  to={{ number: 1 }}>
  {props => <div>{props.number}</div>}
</Spring>`}
    options={{
      mode: 'jsx',
      theme: 'dracula',
      lineNumbers: true,
      readOnly: true,
      showCursorWhenSelecting: false,
      cursorHeight: 0
    }}
  />
  <div>
    <RewindSpring>
      {x => <animated.div style={{ fontFamily: 'monospace'}}>{x.interpolate(n => n.toFixed(2))}</animated.div>}
    </RewindSpring>
  </div>
</div>

<p>
  or generic <span className="highlight">React-component props</span>.
</p>

<div className="code-table">
  <CodeMirror
    value={
`<Spring
  from={{ value: 0 }}
  to={{ value: 100 }}>
  {props => <Donut value={props.value} />}
</Spring>`}
    options={{
      mode: 'jsx',
      theme: 'dracula',
      lineNumbers: true,
      readOnly: true,
      showCursorWhenSelecting: false,
      cursorHeight: 0
    }}
  />
  <div>
    <RewindSpring>
      {x => <AnimatedMeter type="circle" value={x.interpolate(x => x * 100)} />}
    </RewindSpring>
  </div>
</div>
</RewindSpringProvider>

<p>
  Springs are <span className="highlight">physics based</span> and animate adaptively, you do not have to worry about durations and easings, which get complex and illogical quickly. More about that <a href="/gotchas">here</a>.
</p>

## Interpolation

<p>
  <code>Spring</code> doesn't just handle numbers, it can <span className="highlight">interpolate pretty much everything</span>.
</p> 

* colors (names, rgb, rgba, hsl, hsla)
* absolute lenghts (cm, mm, in, px, pt, pc)
* relative lengths (em, ex, ch, rem, vw, vh, vmin, vmax, %)
* angles (deg, rad, grad, turn)
* flex and grid units (fr, etc)
* all HTML attributes
* SVG paths (as long as the number of points matches, otherwise use [custom interpolation](https://codesandbox.io/embed/lwpkp46om))
* arrays
* string patterns (`transform`, `border`, `boxShadow`, etc)
* `auto` is valid
* non-animatable string values (`visibility`, `pointerEvents`, etc)
* `scrollTop`/`scrollLeft` ([native only](/native), since these aren't actual dom properties)

Of course the only properties that the browser can animate relatively cheap are [composite propertys](https://www.html5rocks.com/en/tutorials/speed/high-performance-animations/) (opacity, transform), so you might want to watch out for that.

The following playground is live.

<Playground style={{ height: 200 }}>
  <Spring
    from={{
      width: 100,
      padding: 0,
      background: 'linear-gradient(to right, #30e8bf, #ff8235)',
      transform: 'perspective(600px) translate3d(400px,0,0) scale(2) rotateX(90deg)',
      boxShadow: '0px 100px 150px -10px #2D3747',
      borderBottom: '0px solid white',
      shape: 'M20,380 L380,380 L380,380 L200,20 L20,380 Z',
      textShadow: '0px 5px 0px white'
    }}
    to={{ 
      width: 'auto', 
      padding: 20, 
      background: 'linear-gradient(to right, #009fff, #ec2f4b)', 
      transform: 'perspective(600px) translate3d(0px,0,0) scale(1) rotateX(0deg)', 
      boxShadow: '0px 10px 20px 0px rgba(0,0,0,0.4)', 
      borderBottom: '10px solid #2D3747', 
      shape: 'M20,20 L20,380 L380,380 L380,20 L20,20 Z', 
      textShadow: '0px 5px 15px rgba(255,255,255,0.5)' 
    }} config={{ tension: 10, friction: 50 }}>
    {({ width, padding, background, transform, boxShadow, borderBottom, textShadow, shape }) => (
      <div style={{ display: 'inline-block',
      overflow: 'hidden',
      whiteSpace: 'nowrap',
      fontSize: '4em',
      fontWeight: 800, width, padding, background, transform, boxShadow, textShadow }}>
        <svg style={{ display: 'inline', width: 50 }} viewBox="0 0 400 400">
          <path d={shape} fill="#2D3747"/>
        </svg>
        <span style={{ borderBottom }}>üòÖ</span>
      </div>
    )}
  </Spring>
</Playground>

## Config

Spring are configurable, there are a couple of presets.

```jsx
import { Spring, config } from 'react-spring'

<Spring config={config.default} />
````

| Property                  | Value |
| ------------------------- | ------- |
| config.default             | { tension: 170, friction: 14 }
| config.wobbly              | { tension: 180, friction: 12 }
| config.stiff               | { tension: 210, friction: 20 }
| config.slow                | { tension: 280, friction: 60 }
| config.molasses            | { tension: 280, friction: 120 }

You can of course set configs for yourself, or finetune your spring. 

```jsx
<Spring config={{ tension: 0, friction: 2, restDisplacementThreshold: 0.1 }} />
````

| Property                  | Default | Description       |
| ------------------------- | ----------- | ------------- |
| overshootClamping         | false       | false, when true, stops the spring once it overshoots its boundaries
| restDisplacementThreshold | 0.0001      | precision
| restSpeedThreshold        | 0.0001      | precision
| velocity                  | 0           | initial velocity
| tension                   | 170         | energetic load
| friction                  | 26          | resistence
| delay                     | 0           | delay

It is also possible to set configs key by key.

```jsx
<Spring
  to={{ width: 10, color: 'red' }}
  config={key => key === 'width' ? config.slow : config.wobbly />
```

## Props

<PropsTable of={Spring} />