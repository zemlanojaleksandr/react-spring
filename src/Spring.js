import React from 'react'
import PropTypes from 'prop-types'
import controller from './animated/AnimatedController'
import AnimatedValue from './animated/AnimatedValue'
import AnimatedArray from './animated/AnimatedArray'
import AnimatedProps from './animated/AnimatedProps'
import SpringAnimation from './animated/SpringAnimation'
import * as Globals from './animated/Globals'

function shallowDiff(a, b) {
  for (let i in a) if (!(i in b)) return true
  for (let i in b) if (a[i] !== b[i]) return true
  return false
}

export const config = {
  default: { tension: 170, friction: 26 },
  gentle: { tension: 120, friction: 14 },
  wobbly: { tension: 180, friction: 12 },
  stiff: { tension: 210, friction: 20 },
  slow: { tension: 280, friction: 60 },
}

const callProp = (p, n) => (typeof p === 'function' ? p(n) : p)
const convert = (acc, [name, value]) => ({
  ...acc,
  [name]: new AnimatedValue(value),
})

export default class Spring extends React.Component {
  static propTypes = {
    to: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
    from: PropTypes.object,
    native: PropTypes.bool,
    onStart: PropTypes.func,
    onRest: PropTypes.func,
    onFrame: PropTypes.func,
    children: PropTypes.oneOfType([
      PropTypes.func,
      PropTypes.arrayOf(PropTypes.func),
      PropTypes.node,
    ]),
    render: PropTypes.func,
    reset: PropTypes.bool,
    config: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
    immediate: PropTypes.oneOfType([PropTypes.bool, PropTypes.func]),
    impl: PropTypes.func,
    inject: PropTypes.func,
  }

  static defaultProps = {
    from: {},
    to: {},
    config: config.default,
    native: false,
    immediate: false,
    reset: false,
    impl: SpringAnimation,
    inject: Globals.bugfixes,
  }

  state = { props: undefined }
  animations = {}

  componentWillUnmount() {
    this.stop()
  }

  componentWillMount() {
    this.updatePropsAsync(this.props)
  }

  componentWillUpdate(props) {
    if (props.reset || shallowDiff(props.to, this.props.to))
      this.updatePropsAsync(props)
  }

  updatePropsAsync(props) {
    if (props.inject) {
      this.inject = props.inject(this, props)
      if (this.inject) return
    }
    this.updateProps(props)
  }

  updateProps(props, force = false, didInject = false) {
    // Springs can be destroyed, the "destroyed" flag prevents them from ever
    // updating further, they'll just animate out and function no more ...
    if (this.destroyed && props.destroyed) return
    this.destroyed = props.destroyed

    const {
      impl,
      from,
      to,
      config,
      attach,
      immediate,
      reset,
      onFrame,
      onRest,
      inject,
      native,
    } = props
    const allProps = Object.entries({ ...from, ...to })

    this.interpolators = {}
    this.animations = allProps.reduce((acc, [name, value], i) => {
      const entry =
        (reset === false && this.animations[name]) ||
        (this.animations[name] = {})

      let isNumber = typeof value === 'number'
      let isString =
        typeof value === 'string' &&
        !value.startsWith('#') &&
        !/\d/.test(value) &&
        !Globals.colorNames[value]
      let isArray = !isNumber && !isString && Array.isArray(value)
      let fromValue = from[name] !== undefined ? from[name] : value
      let fromAnimated = fromValue instanceof AnimatedValue
      let toValue = isNumber || isArray ? value : 1

      if (/*isNumber &&*/ attach) {
        // Attach value to target animation
        const target = attach(this)
        const targetAnimation = target && target.animations[name]
        if (targetAnimation) toValue = targetAnimation.animation
      }

      if (fromAnimated) {
        // Use provided animated value
        entry.animation = entry.interpolation = fromValue
      } else if (isNumber || isString) {
        // Create animated value
        entry.animation = entry.interpolation =
          entry.animation || new AnimatedValue(fromValue)
      } else if (isArray) {
        // Create animated array
        entry.animation = entry.interpolation =
          entry.animation || new AnimatedArray(fromValue)
      } else {
        // Deal with interpolations
        const previous =
          entry.interpolation &&
          entry.interpolation._interpolation(entry.animation._value)
        entry.animation = new AnimatedValue(0)
        entry.interpolation = entry.animation.interpolate({
          range: [0, 1],
          output: [previous !== undefined ? previous : fromValue, value],
        })
      }

      if (callProp(immediate, name)) entry.animation.setValue(toValue)

      entry.stopped = false

      entry.onFinish = cb => {
        this.animations[name].stopped = true
        if (this.getAnimations().every(a => a.stopped)) {
          const current = { ...this.props.from, ...this.props.to }
          if (onRest) onRest(current)
          cb && typeof cb === 'function' && cb(current)

          if (didInject) {
            // Restore the original values for injected props
            const componentProps = this.convertValues(this.props)
            this.inject = this.renderChildren(this.props, componentProps)
            this.forceUpdate()
          }
        }
      }

      entry.start = cb => {
        if (entry.animation.__getValue() === toValue) return entry.onFinish(cb)
        controller(
          entry.animation,
          { to: toValue, ...callProp(config, name) },
          impl
        ).start(props => props.finished && entry.onFinish(cb))
      }

      entry.stop = () => {
        entry.stopped = true
        entry.animation.stopAnimation()
      }

      this.interpolators[name] = entry.interpolation
      return { ...acc, [name]: entry }
    }, {})

    const oldAnimatedProps = this.animatedProps
    this.animatedProps = new AnimatedProps(this.interpolators, this.callback)
    oldAnimatedProps && oldAnimatedProps.__detach()

    this.updateToken = true
    if (force) this.forceUpdate()
  }

  start() {
    const { onStart } = this.props

    let fn = () =>
      this.getAnimations().forEach(animation => animation.start(resolve))
    let resolve,
      promise = new Promise(r => (resolve = r))
    if (this.props.delay) {
      if (this.timeout) clearTimeout(this.timeout)
      return (this.timeout = setTimeout(() => fn(), this.props.delay))
    }

    if (onStart) onStart()

    fn()
    return promise
  }

  stop() {
    this.getAnimations().forEach(animation => animation.stop())
  }

  flush() {
    this.getAnimations().forEach(
      ({ interpolation }) => interpolation._update && interpolation._update()
    )
  }

  callback = () => {
    if (this.props.onFrame) this.props.onFrame(this.animatedProps.__getValue())
    !this.props.native && this.forceUpdate()
  }

  getAnimations() {
    return Object.keys(this.animations).map(key => this.animations[key])
  }

  getValues() {
    return this.animatedProps ? this.animatedProps.__getValue() : {}
  }

  getAnimatedValues() {
    return this.props.native ? this.interpolators : this.getValues()
  }

  convertValues(props) {
    const { from, to, native, children, render } = props
    const forward = this.getForwardProps(props)
    const allProps = Object.entries({ ...from, ...to })
    return native
      ? allProps.reduce(convert, forward)
      : { ...from, ...to, ...forward }
  }

  getForwardProps(props = this.props) {
    const {
      to,
      from,
      config,
      native,
      onRest,
      onFrame,
      children,
      render,
      reset,
      immediate,
      impl,
      inject,
      ...forward
    } = props
    return forward
  }

  componentDidUpdate() {
    // Animation has to start *after* render, since at that point the scene
    // graph should be established, so we do it here. Unfortunatelly, non-native
    // animations call forceUpdate, so it's causing a loop. updateToken prevents
    // that as it gets set only on prop changes.
    if (this.updateToken) {
      this.updateToken = false
      this.start()
    }
  }

  componentDidMount() {
    this.updateToken = false
    this.start()
  }

  renderChildren(props, componentProps) {
    return props.render
      ? props.render({ ...componentProps, children: props.children })
      : props.children(componentProps)
  }

  render() {
    if (this.inject) {
      const content = this.inject
      this.inject = undefined
      return content
    }

    const { children, render } = this.props
    const values = this.getAnimatedValues()
    return values && Object.keys(values).length
      ? this.renderChildren(this.props, {
          ...values,
          ...this.getForwardProps(),
        })
      : null
  }
}
