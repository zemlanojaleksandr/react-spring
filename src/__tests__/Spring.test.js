import React from 'react'
import { shallow } from 'enzyme'
import Spring from '../Spring'

const fakeChild = () => <div>fake</div>

// These snapshots won't reveal much, but they at least will catch hard errors.
describe('Spring', () => {
  it('smoke/regression tests', () => {
    const spring = shallow(<Spring />)
    expect(spring).toMatchSnapshot()
  })

  it('fires componentDidUpdate on componentDidMount', () => {
    const spy = jest.spyOn(Spring.prototype, 'componentDidUpdate')
    const spring = shallow(<Spring />)
    spring.instance().componentDidMount()
    expect(spy).toHaveBeenCalled()
    spy.mockRestore()
  })
})
