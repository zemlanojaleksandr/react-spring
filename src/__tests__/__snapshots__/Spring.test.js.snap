// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spring smoke/regression tests 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Spring
    config={
      Object {
        "friction": 26,
        "tension": 170,
      }
    }
    force={false}
    from={Object {}}
    immediate={false}
    native={false}
    reset={false}
    to={Object {}}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): null,
  Symbol(enzyme.__nodes__): Array [
    null,
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
